Implement util function compute_global_tensor_shape for 1D device mesh

dharakk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #152166
* __->__ #151990

### Summary

compute_global_tensor_shape util function takes in local tensor shape, device mesh
and placements. We all gather the shapes from the shards and according to the placement
type we construct the global shape.

Note: currenty only implemented for placement type Shard and Replicate, TODO for StridedShared

### Test

`pytest test/distributed/tensor/test_utils.py`

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

Commit:
Implement util function compute_global_tensor_shape for 1D device mesh

Summary:

compute_global_tensor_shape util function takes in local tensor shape, device mesh
and placements. We all gather the shapes from the shards and according to the placement
type we construct the global shape.

Note: currenty only implemented for placement type Shard and Replicate, TODO for StridedShared

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "Implement util function compute_global_tensor_shape for 1D device mesh"


Summary:

compute_global_tensor_shape util function takes in local tensor shape, device mesh
and placements. We all gather the shapes from the shards and according to the placement
type we construct the global shape.

Note: currenty only implemented for placement type Shard and Replicate, TODO for StridedShared

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

dharakk committed:
Update on "Implement util function compute_global_tensor_shape for 1D device mesh"


compute_global_tensor_shape util function takes in local tensor shape, device mesh
and placements. We all gather the shapes from the shards and according to the placement
type we construct the global shape.

Note: currenty only implemented for placement type Shard and Replicate, TODO for StridedShared


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

XilunWu commented on torch/distributed/tensor/_utils.py:
```diff
@@ -246,6 +246,58 @@ def compute_global_tensor_info(
+    shape: torch.Size, mesh: DeviceMesh, placements: Sequence[Placement]
+) -> torch.Size:
+    """
+    Compute the global size of a DTensor from the given local tensor shape,
+    the mesh and placements.
```
let's elaborate on how this function is different from `compute_global_tensor_info`: this util collects the tensor shard shapes instead of assuming that sharding is even, and the reason we need this util is we want to have a way to calculate the full tensor shape from shards' shapes.

XilunWu commented on torch/distributed/tensor/_utils.py:
```diff
@@ -246,6 +246,58 @@ def compute_global_tensor_info(
+
+    Args:
+        shape (:class:`torch.Size`):
+            Shape of the Local tensor
```
Local -> local

XilunWu commented on torch/distributed/tensor/_utils.py:
```diff
@@ -246,6 +246,58 @@ def compute_global_tensor_info(
+            on the mesh topology.
+
+    Return:
+        tensor_shape: Shape of the glocal DTensor.
```
glocal -> global

XilunWu commented on torch/distributed/tensor/_utils.py:
```diff
@@ -246,6 +246,58 @@ def compute_global_tensor_info(
+    elif isinstance(placements[0], Shard):
+        local_shape = torch.tensor(list(shape))
+        gathered_shapes = [None for _ in range(torch.distributed.get_world_size())]
+        torch.distributed.all_gather_object(gathered_shapes, local_shape)
```
`all_gather_object` is for non-Tensor objects, you can use `all_gather` here.

XilunWu commented on test/distributed/tensor/test_utils.py:
```diff
@@ -141,6 +142,39 @@ def _compute_start_end_offsets(self, global_offset, local_size, n_dim):
 
+    @with_comms
+    def test_compute_global_tensor_shape_1D(self):
+        one_d_placements = [[Shard(1)], [Replicate()]]
```
we can also add `[Shard(0)]` to the list

XilunWu reviewed (commented):
Overall LGTM. Left some suggestions. Next step is to extend to multi-dims device mesh.

dharakk committed:
Update on "Implement util function compute_global_tensor_shape for 1D device mesh"


compute_global_tensor_shape util function takes in local tensor shape, device mesh
and placements. We all gather the shapes from the shards and according to the placement
type we construct the global shape.

Note: currenty only implemented for placement type Shard and Replicate, TODO for StridedShared


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

dharakk committed:
Update on "Implement util function compute_global_tensor_shape for 1D device mesh"


compute_global_tensor_shape util function takes in local tensor shape, device mesh
and placements. We all gather the shapes from the shards and according to the placement
type we construct the global shape.

Note: currenty only implemented for placement type Shard and Replicate, TODO for StridedShared


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

XilunWu commented on test/distributed/tensor/test_utils.py:
```diff
@@ -141,6 +142,52 @@ def _compute_start_end_offsets(self, global_offset, local_size, n_dim):
+                    if placements[0].dim == 1
+                    else torch.Size([sum(uneven_dim), 5])
+                )
+                global_shape = compute_global_tensor_shape(
+                    local_shape,
+                    device_mesh,
+                    placements,
+                )
+                self.assertEqual(
+                    global_shape,
+                    expected_global_shape,
+                )
+            else:
+                global_shape = compute_global_tensor_shape(
+                    torch.Size([5, 5]), device_mesh, placements
+                )
+                self.assertEqual(global_shape, torch.Size([5, 5]))
```
this common piece of code can be moved out of the `if-else` with minor change. We can just calculate `local_shape` and `expected` in the branch.

XilunWu commented on torch/distributed/tensor/_utils.py:
```diff
@@ -246,6 +246,60 @@ def compute_global_tensor_info(
+        raise NotImplementedError(
+            "compute_global_tensor_shape only supports 1D mesh for now."
+        )
+    if len(placements) != 1:
+        raise NotImplementedError(
+            "compute_global_tensor_shape only supports 1 placement for now."
+        )
```
we can actually check `len(placements) == mesh.ndim` instead which is also a hard requirement.

XilunWu commented on torch/distributed/tensor/_utils.py:
```diff
@@ -246,6 +246,60 @@ def compute_global_tensor_info(
+        local_shape = torch.tensor(list(shape))
+        gathered_shaped_tensors = [
+            torch.empty_like(local_shape, device=local_shape.device)
+            for _ in range(torch.distributed.get_world_size())
```
`torch.distributed.get_world_size()` -> `self.world_size`

dharakk replied:
Here we would not have a `self` object, may be can use `mesh.size()` ?

XilunWu replied:
Right. This is the right value to use. Think about a world that's separated into multiple parallelism axes (e.g. FSDP + TP), then `WORLD_SIZE` != `mesh.size()` if DTensor is sharded over 1D mesh (e.g. TP).

XilunWu commented on torch/distributed/tensor/_utils.py:
```diff
@@ -246,6 +246,60 @@ def compute_global_tensor_info(
+            torch.empty_like(local_shape, device=local_shape.device)
+            for _ in range(torch.distributed.get_world_size())
+        ]
+        torch.distributed.all_gather(gathered_shaped_tensors, local_shape)
```
let's use the functional collectives by:
```
# at file head
import torch.distributed._functional_collectives as funcol

# in code
funcol.all_gather_tensor(...)
```

see file `torch/distributed/_functional_collectives.py` for function reference and `torch/distributed/tensor/placement_types.py` for example usage.

dharakk replied:
can we use `funcol.all_gather_inplace(...)` https://fburl.com/x2xi8mhc which has the same interface taking in `List[Tensor]` and a `Tensor`

XilunWu replied:
These 2 options are pretty the same in this case:
1) in order to `all_gather_inplace`, you need to first create the output list.
2) after calling `all_gather_tensor`, you need to `torch.chunk` to create a tensor list.

it's only a matter of ordering.

XilunWu commented on torch/distributed/tensor/_utils.py:
```diff
@@ -246,6 +246,60 @@ def compute_global_tensor_info(
+            for _ in range(torch.distributed.get_world_size())
+        ]
+        torch.distributed.all_gather(gathered_shaped_tensors, local_shape)
+        sharded_dim_sum = 0
+        for shape_tensor in gathered_shaped_tensors:
+            shape_tensor_list = shape_tensor.tolist()
+            sharded_dim_sum += shape_tensor_list[placements[0].dim]
```
```
shard_dim = placements[0].dim
other_dims = [d for d in range(shape_tensor.ndim) if d != shard_dim]
sharded_dim_size = shape_tensor[shard_dim].item()
other_dim_size = shape_tensor[other_dims]
```

then sum up `sharded_dim_size`, check if `other_dim_size` are consistent.

XilunWu reviewed (commented):
Thanks for addressing the feedbacks! Added some more review feedbacks. Besides, let's add `Test` section in the PR description as well.

dharakk committed:
Update on "Implement util function compute_global_tensor_shape for 1D device mesh"


compute_global_tensor_shape util function takes in local tensor shape, device mesh
and placements. We all gather the shapes from the shards and according to the placement
type we construct the global shape.

Note: currenty only implemented for placement type Shard and Replicate, TODO for StridedShared


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

XilunWu reviewed (approved):
Thanks for the PR!

dharakk commented:
@pytorchbot  merge -f "test_dynamic_sources_dynamic_override failing which is unrelated to this change"

dharakk commented:
Recreating PR #152751 to mitigate easySLA failure
